### ISSUE TITLE
Please review changes to falk-with-tests branch

### ISSUE THREAD

**Issue by postfalk** [tnc-ca-geo/animl-email-relay#11]
Hi @nathanielrindlaub,

Before I create a pull request I would like you to functional test the reworked module in the dev environment. I also would be happy to discuss and answer questions. The test coverage is not quite 100% yet but it would take me to much time to make it picture perfect, it is more about the principle and writing code for testibility as a major design concern. 

Note, I entirely rewrote the code without ever test it in its natural environment. In a way that is the purpose of tests. So what I did is, 1. wrote tests around the handler to make sure that it still does the same thing after changes. 2. isolating functional units, 3. rewriting those, 4. write tests for the new units and test whether the overall module still does the same.

BUT of course, I cannot be sure that my testing hits all possible use cases.

For 2. Isolating functional units. There are a few criteria that are important. On the top level, I wanted to write the handler in a way that reads like pseudocode to make very clear what it does. Then there are a lot of little things that are in a way generic: reading and parsing emails, writing exif, etc. those were reworked in generic functions. It is also a very good idea to pack all side effects in rather short functions, that can be easy mocked (instead of mocking the functionality of third party libraries, which has been done in places as well, e.g. requests). In general tests should not relay on specific resources, API, logins, environmental variables etc. In the case of exiftool, I made an exemption since it is a very central functionality and changes in Exiftool could possible break the whole thing. In sofar it is sensible to test against a certain version (there are test runners that can actually do that, but I don't).

**Comment by nathanielrindlaub** [comment]
@postfalk, quick update on this - I deployed your branch to dev and tested it and it works perfectly. Thank you!! I would say we're ready to merge it if you want to create a PR. 

I reviewed the re-worked code structure and love it; although I haven't reviewed the test-related code yet. I'll have a look before our next check in though.

Thank you for your work on this - it is looking MUCH improved.
